### PR TITLE
Iteration 5

### PR DESCRIPTION
**Pull request by gratefulcheddar**
Redesigned most of the code so that the different game modes are easier to implement.

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -1,20 +1,76 @@
...
   def initialize
-    @history = []
+    @turns = []
+    @secret_code = nil
   end
 
-  def add(guess)
-    @history << guess
+  def add_secret(code)
+    @secret_code = code
```
Comment (ssunday) [inline lib/board.rb]: This doesn't look like it's used in this class at all.

**Comment by gratefulcheddar** [comment]
Closing this because it looks like some issues came up when merging my changes on top of master...

File: lib/game_controller.rb [hunk 2]
```diff
@@ -1,41 +1,11 @@
...
-      end
-
-      if turn_number == Mastermind::MAX_TURNS
-        @io.output @messages.out_of_turns_message(@game.secret_code)
-        break
-      end
-
-    end
+  def self.launch_game
+    io = MastermindIO.new
```
Comment (ssunday) [inline lib/game_controller.rb]: It would be nice if these classes were injected in...that would be a good next step

File: lib/board.rb [hunk 3]
```diff
@@ -1,20 +1,76 @@
...
-    board
+    board_string << "\n"
+    board_string
+  end
+end
+
+class Turn
+  attr_accessor :feedback
+  attr_reader :guess
+
```
Comment (ssunday) [inline lib/board.rb]: I like this, but maybe pull it out into another file?

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    end
+    @board.add_guess(code)
+  end
+
+  private 
+
+  def calculate_new_combinations
+    previous_guess_pins = @board.last_pins
+    new_combinations = []
+    @possible_combinations.each do |combination|
```
Comment (ssunday) [inline lib/player.rb]: Can't new combinations just be a map of this iteration?

File: lib/rules.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module Rules
+  COLOR_OPTIONS = %i[red blue green orange purple yellow].freeze
+  GAME_OPTIONS = %i[quit restart].freeze
+  GAME_MODES = %i[computer human auto].freeze
+  SECRET_LENGTH = 4
```
Comment (ssunday) [inline lib/rules.rb]: I like this

File: spec/board_spec.rb [hunk 6]
```diff
@@ -4,21 +4,108 @@
...
+  end
+
+  describe '#add_guess(guess)' do
+    it 'adds a turn to the board with specified guess and no feedback' do
+      code = %i[red red red red]
+      expect(game_board.turn_count).to eq 0
+      game_board.add_guess(code)
+      expect(game_board.turn_count).to eq 1
+      expect(game_board.last_guess).to eq code
+      expect(game_board.last_pins).to eq []
```
Comment (ssunday) [inline spec/board_spec.rb]: to be_empty is a neat shortcut, be_zero as well

**Pull request closed without merging**